core/cost-model.md

```diff
@@ -0,0 +1,153 @@
+  Cost of built-in instructions
+
+- **NONBUILTIN_INSTR_COST**: 200000  
+  Cost of non-builtin instructions
+
+- **MAX_COMPUTE_UNIT_LIMIT**: 1400000  
+  Maximum compute unit limit
+
+- **HEAP_PAGE_COST**: 8  
+  Cost per heap page
+
+- **HEAP_PAGE_SIZE**: 32768  
+  Size of one heap page (32KB)
+
+- **MAX_LOADED_ACCT_DATA_SZ**: 64 * 1024 * 1024
+  Default size (64MB)
+
+- **SIMPLE_VOTE_COST**: 3428  
+  Fixed cost for simple vote transactions
+
+## Messages
+
+### SignatureCost
+
+`SignatureCost` is the cost due to signature verification via builtin verifiers. This cost can be computed deterministically from the transaction payload.
+
+**`SignatureCost` is computed as:**
+```
+(TXN_SIGNATURE_COST * txn_sig_count) +
+(ED25519_SIGNATURE_COST * ed25519_sig_count) +
+(SECP256R1_SIGNATURE_COST * secp256r1_sig_count)
+```
+
+- `txn_sig_count`: the native signatures included directly int the transaction
+- `ed25519_sig_count`: the number of signatures used by ed25519 precompiles
+- `secp256r1_sig_count`: the number of signatures used by secp256r1 precompiles
+
+
+### WriteCost
+
+`WriteCost` is the cost due to having to acquire write-locks on accounts. This cost can be computed deterministically from the transaction payload.
+
+**`WriteCost` is computed as:**
+```
+WRITABLE_ACCT_COST * num_writable_accts
+```
+
+- `num_writable_accts`: The number of accounts in a transaction that are flagged as writable in the transaction payload.
+
+### InstructionDataCost
+
+`InstructionDataCost` is the cost incurred by the instruction data in the transaction. This cost can be computed deterministically from the transaction payload.
+
+**InstructionDataCost is computed as:**  
+```
+INSTR_DATA_COST * instr_data_sz
+```
+
+- `instr_data_sz`: Sum of instruction data sizes for all instructions included in the transaction.
+
+### EstimatedExecutionCost
+
+`EstimatedExecutionCost` is the estimated cost for instruction execution. This estimate is provided by a `ComputeBudgetInstruction::SetComputeUnitLimit(u32)` instruction. If this instruction is not present, then default cost values are used. If it can be determined that a transaction will fail before it is executed, then its associated total_requested_execution_cost is zero (for fee-only transactions). This cost can be computed deterministically from the transaction payload.
+
+**EstimatedExecutionCost is determined as:**  
+
+```
+compute_instr_is_provided
+    ? MIN(instruction_requested_cost, MAX_COMPUTE_UNIT_LIMIT)
+    : (BUILTIN_INSTR_COST * num_builtin_instrs) +
+      (NONBUILTIN_INSTR_COST * num_nonbuiltin_instrs)
+```
+
+- `compute_instr_is_provided`: True iff the transaction includes a `ComputeBudgetInstruction::SetComputeUnitLimit(u32)` instruction
+
+- `instruction_requested_cost`: The requested compute units provided by `ComputeBudgetInstruction::SetComputeUnitLimit(u32)`.
+
+- `num_builtin_instrs`: The number of builtin instructions. A list of builtin instructions was introduced by SIMD 0170.
```

Comment:
nit: SIMD-170 defined the const MIN CUs for builtin, "what are builtins" is defined [here](https://github.com/anza-xyz/agave/blob/master/builtins-default-costs/src/lib.rs), again, some are based on migration feature gate status.